File: hsl_ma97-2.3.0/example/Fortran/hsl_ma97ds1.f90
! Simple code to illustrate use of hsl_ma97
program hsl_ma97ds1
   use hsl_ma97_double
   implicit none

   ! Derived types
   type (ma97_akeep)   :: akeep
   type (ma97_fkeep)   :: fkeep
   type (ma97_control) :: control
   type (ma97_info)    :: info

   ! Parameters
   integer, parameter :: wp = kind(0.0d0)

   integer, dimension (:),  allocatable :: row
   integer, dimension (:),  allocatable :: col
   real(wp), dimension (:), allocatable :: val
   real(wp), dimension (:,:), allocatable :: x

   integer :: matrix_type,n,ne

   ! Read in the order n of the matrix and number of entries in lower triangle
   read (*,*) n,ne

   ! Allocate arrays for matrix data and arrays for hsl_ma97
   allocate (row(ne),col(ne),val(ne))
   allocate (x(n,2))

   read (*,*) row(1:ne)
   read (*,*) col(1:ne)
   read (*,*) val(1:ne)

   ! Perform analyse and factorise with coordinate input
   call ma97_analyse_coord(n,ne,row,col,akeep,control,info)
   if (info%flag < 0) go to 100
   matrix_type = 4 ! Real, symmetric indefinite
   call ma97_factor(matrix_type,val,akeep,fkeep,control,info)
   if (info%flag < 0) go to 100

   ! Read in the right-hand side
   read (*,*) x(1:n,1:2)

   ! Solve
   call ma97_solve(2,x,n,akeep,fkeep,control,info)
   if (info%flag < 0) go to 100
   write (*,'(/a,/,(3es18.10))') ' The computed solution is:', x(1:n,1)
   write (*,'(3es18.10)') x(1:n,2)

   ! Read values of second matrix with same pattern
   read (*,*) val(1:ne)

   ! Read another right hand side
   read (*,*) x(1:n,1)

   ! Perform combined factor and solve
   call ma97_factor_solve(matrix_type,val,x(1:n,1),akeep,fkeep,control,info)
   write (*,'(/a,/,(3es18.10))') ' Next solution is:', x(1:n,1)

   100 continue
   call ma97_finalise(akeep, fkeep)

end program hsl_ma97ds1

File: hsl_ma97-2.3.0/hsl_ma97-2.3.0/example/C/hsl_ma97ds1.c
/* hsl_ma97ds1.c */
/* Simple code to illustrate coordinate entry for hsl_ma97 */

#include <stdio.h>
#include <stdlib.h>
#include "hsl_ma97d.h"

int main(void) {
   typedef double pkgtype;
   
   void *akeep, *fkeep;
   struct ma97_control control;
   struct ma97_info info;

   int *row, *col;
   pkgtype *val, *x;

   int i,matrix_type,n,ne;

   /* Read in the order n of the matrix and number of entries in lwr triangle */
   scanf("%d %d", &n, &ne);

   /* Allocate arrays for matrix data and arrays for hsl_ma97 */
   row = (int *) malloc(ne*sizeof(int));
   col = (int *) malloc(ne*sizeof(int));
   val = (pkgtype *) malloc(ne*sizeof(pkgtype));
   x = (pkgtype *) malloc(2*n*sizeof(pkgtype));

   for(i=0; i<ne; i++) scanf("%d", &(row[i]));
   for(i=0; i<ne; i++) scanf("%d", &(col[i]));
   for(i=0; i<ne; i++) scanf("%lf", &(val[i]));

   ma97_default_control(&control);

   /* Perform analyse and factorise with data checking */
   ma97_analyse_coord(n,ne,row,col,NULL,&akeep,&control,&info,NULL);
   if (info.flag < 0) {
      ma97_free_akeep(&akeep);
      free(row); free(col); free(val); free(x);
      return 1;
   }
   matrix_type = 4; /* Real, symmetric indefinite */
   fkeep = NULL; /* important that this is initialised to NULL on first call */
   ma97_factor(matrix_type,NULL,NULL,val,&akeep,&fkeep,&control,&info,NULL);
   if (info.flag < 0) {
      ma97_finalise(&akeep,&fkeep);
      free(row); free(col); free(val); free(x);
      return 1;
   }

   /* Read in the right-hand sides. */
   for(i=0; i<2*n; i++) scanf("%lf", &(x[i]));

   /* Solve */
   ma97_solve(0,2,x,n,&akeep,&fkeep,&control,&info);
   if (info.flag < 0) {
      ma97_finalise(&akeep,&fkeep);
      free(row); free(col); free(val); free(x);
      return 1;
   }
   printf("\nThe computed solution is:\n");
   for(i=0; i<n; i++) printf(" %lf", x[i]);
   printf("\n");
   for(i=0; i<n; i++) printf(" %lf", x[n+i]);

   /* Read second matrix with same pattern */
   for(i=0; i<ne; i++) scanf("%lf", &(val[i]));

   /* Read another right hand side */
   for(i=0; i<n; i++) scanf("%lf", &(x[i]));

   /* Perform combined factor and solve */
   /* Note: no need to set fkeep to NULL, we will overwrite existing factors */
   ma97_factor_solve(matrix_type,NULL,NULL,val,1,x,n,&akeep,&fkeep,&control,
         &info,NULL);
   printf("\nNext solution is:\n");
   for(i=0; i<n; i++) printf(" %lf", x[i]);
   printf("\n");

   ma97_finalise(&akeep,&fkeep);

   /* Deallocate all arrays */
   free(row); free(col); free(val); free(x);

   return 0;
}
